irc/src/connection/tls.rs

```diff
@@ -0,0 +1,131 @@
+use std::{io::Cursor, path::PathBuf, sync::Arc};
+
+use bytes::Bytes;
+use tokio::{fs, net::TcpStream};
+use tokio_rustls::{
+    client::TlsStream,
+    rustls::{
+        self,
+        client::danger::{self, ServerCertVerifier},
+        pki_types,
+    },
+    TlsConnector,
+};
+
+pub async fn connect<'a>(
+    tcp: TcpStream,
+    server: &str,
+    accept_invalid_certs: bool,
+    root_cert_path: Option<&'a PathBuf>,
+    client_cert_path: Option<&'a PathBuf>,
+    client_key_path: Option<&'a PathBuf>,
+) -> Result<TlsStream<TcpStream>, Error> {
+    let mut roots = rustls::RootCertStore::empty();
+
+    for cert in rustls_native_certs::load_native_certs()? {
+        roots.add(cert).unwrap();
+    }
+    if let Some(cert_path) = root_cert_path {
+        let cert_bytes = fs::read(&cert_path).await?;
+        let certs =
+            rustls_pemfile::certs(&mut Cursor::new(&cert_bytes)).collect::<Result<Vec<_>, _>>()?;
+        roots.add_parsable_certificates(certs);
+    }
```

Comment:
This is only relevant if `accept_invalid_certs == false`

Reply:
Good point 364db2e